kernel/src/transaction/mod.rs

```diff
@@ -271,51 +317,114 @@ impl Transaction {
+    /// Convert file metadata provided by the engine into protocol-compliant add actions.
+    fn generate_adds<'a, I, T>(
+        &'a self,
+        engine: &dyn Engine,
+        add_files_metadata: I,
+        input_schema: SchemaRef,
+        output_schema: SchemaRef,
+    ) -> EngineDataIterator<'a>
+    where
+        I: Iterator<Item = DeltaResult<T>> + Send + 'a,
+        T: Deref<Target = dyn EngineData> + Send + 'a,
+    {
+        let evaluation_handler = engine.evaluation_handler();
+
+        Box::new(add_files_metadata.map(move |add_files_batch| {
+            // Convert stats to a JSON string and nest the add action in a top-level struct
+            let adds_expr = Expression::struct_from([Expression::transform(
+                Transform::new_top_level().with_replaced_field(
+                    "stats",
+                    Expression::unary(ToJson, Expression::column(["stats"])).into(),
+                ),
+            )]);
+            let adds_evaluator = evaluation_handler.new_expression_evaluator(
+                input_schema.clone(),
+                Arc::new(adds_expr),
+                output_schema.clone().into(),
+            );
+            adds_evaluator.evaluate(add_files_batch?.deref())
+        }))
+    }
 
-        // Step 2: Convert stats to JSON string
-        let field_expressions: Vec<Expression> = intermediate_schema
-            .fields()
-            .map(|f| match f.name().as_str() {
-                "stats" => {
-                    Expression::unary(UnaryExpressionOp::ToJson, Expression::column([f.name()]))
-                }
-                _ => Expression::column([f.name()]),
-            })
-            .collect();
-        let adds_expr = Expression::struct_from([Expression::struct_from(field_expressions)]);
-        let adds_evaluator = evaluation_handler.new_expression_evaluator(
-            intermediate_schema.clone(),
-            Arc::new(adds_expr),
-            log_schema.clone().into(),
+    /// Extend file metadata provided by the engine with row tracking information and convert them into
+    /// protocol-compliant add actions.
+    fn generate_adds_with_row_tracking<'a>(
+        &'a self,
+        engine: &dyn Engine,
+        commit_version: u64,
+    ) -> DeltaResult<EngineDataIterator<'a>> {
+        // Return early if we have nothing to add
+        if self.add_files_metadata.is_empty() {
+            return Ok(Box::new(iter::empty()));
+        }
+
+        // Read the current rowIdHighWaterMark from the snapshot's row tracking domain
+        let row_id_high_water_mark = domain_metadata_configuration(
+            self.read_snapshot.log_segment(),
+            ROW_TRACKING_DOMAIN_NAME,
+            engine,
+        )?
+        .map(|domain_metadata| serde_json::from_str::<RowTrackingDomainMetadata>(&domain_metadata))
+        .transpose()?
+        .and_then(|metadata| metadata.row_id_high_water_mark);
```

Comment:
nit: this could be a static helper method in row_tracking.rs. same with creating a row tracking DomainMetadata action